### PR TITLE
Add receipt scanner online demo

### PR DESCRIPTION
**Pull request by danschultzer**
This adds an online demo.

### CODE REVIEW

File: controllers/api_receipt.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    filename: function (req, file, cb) {
+      crypto.pseudoRandomBytes(16, function (err, raw) {
+        cb(null, err ? undefined : (raw.toString('hex') + '.' + mime.extension(file.mimetype)));
+      });
+    }
+  }),
+  upload = multer({
+    storage: storage,
+    fileFilter: function (req, file, cb) {
+      if (!file.originalname.match(/\.(jpg|jpeg|png|pdf)$/)) {
```
Comment (Schultzer) [inline controllers/api_receipt.js]: What about BMP and TIFF?

File: controllers/api_receipt.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+   * @apiGroup Receipt
+   *
+   * @apiParam {File} document Image or PDF document
+   *
+   * @apiSuccess {Boolean} success
+   * @apiError {String} error Message explaining what went wrong.
+   */
+  router.post('/receipt', upload.single('document'), function(req, res, next) {
+    if (!req.file) {
+      throw new Error("Missing file attachment");
```
Comment (Schultzer) [inline controllers/api_receipt.js]: Use single quotation marks

File: controllers/api_receipt.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    console.log(req.file.path);
+
+    scanner(req.file.path)
+      .parse(function(error, results) {
+        if (error)
+          return next(error);
+
+        res.send({ "success": true, "results": results });
```
Comment (Schultzer) [inline controllers/api_receipt.js:50]: Use single quotation marks

File: public/css/filelist.css [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+  font-family: system, -apple-system, ".SFNSDisplay-Regular", "Helvetica Neue", Helvetica, "Segoe UI", sans-serif;
+  font-size: 13px;
+  line-height: 1.6;
+  color: #333;
+  background-color: transparent;
+}
+
+.window .table {
+  margin: 0px;
```
Comment (Schultzer) [inline public/css/filelist.css]: should just be `0;`

File: public/css/filelist.css [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+  padding: 0px;
```
Comment (Schultzer) [inline public/css/filelist.css]: should just be `0;`

File: public/css/filelist.css [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+  height: 46px;
+  overflow: none;
+}
+
+.window .table .head .cell {
+  color: var(--submenu-black-text);
+}
+
+.window .table .head header.titlebar {
+  margin-bottom: 0px;
```
Comment (Schultzer) [inline public/css/filelist.css]: should just be `0;`

File: public/css/filelist.css [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+.window .head .row {
+  background-color: var(--submenu-bg);
+  border-bottom: 1px solid var(--submenu-border-buttom);
+}
+
+.window .table .body {
+  position: fixed;
+  position:absolute;
+  top: 46px;
+  left: 0px;
```
Comment (Schultzer) [inline public/css/filelist.css]: should just be `0;`

File: public/css/filelist.css [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+  right: 0px;
```
Comment (Schultzer) [inline public/css/filelist.css]: should just be `0;`

File: public/css/filelist.css [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+  bottom: 0px;
```
Comment (Schultzer) [inline public/css/filelist.css]: should just be `0;`

File: public/javascripts/receipt.js [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+    document.body.querySelector('#try-demo').addEventListener('click', function(event) {
+      event.preventDefault();
+      SmoothScrolling.eventHandler(event);
+    });
+
+    document.body.querySelectorAll('form[data-remote]').forEach(function(form) {
+      var submit = function() {
+        if (form.getAttribute('data-submitting')) {
+          console.log("Form is already being submitted");
```
Comment (Schultzer) [inline public/javascripts/receipt.js]: Use single quotation marks

File: public/javascripts/receipt.js [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+            };
+          files.splice(0, 1, file);
+          var interval = setInterval(function(){
+            if (file.progressBar < 50)
+              file.progressBar += Math.random() * (5 - 0) + 0;
+            else
+              clearInterval(interval);
+          }, 200);
+          request.onreadystatechange = function() {
+            if (request.readyState == request.DONE) {
```
Comment (Schultzer) [inline public/javascripts/receipt.js]: this should be `===`

File: public/javascripts/receipt.js [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+                  file.result.parsed = json.results;
+                } else {
+                  file.result.error = json.error;
+                }
+              } catch (error) {
+                alert(error.message);
+              }
+            }
+          };
+          request.open("POST", form.action);
```
Comment (Schultzer) [inline public/javascripts/receipt.js]: Use single quotation marks

File: public/javascripts/receipt.js [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+      });
+
+      form.addEventListener('submit', function(event) {
+        event.preventDefault();
+        submit();
+      });
+    });
+  }
+
+  if (["complete", "loaded", "interactive"].indexOf(document.readyState)) {
```
Comment (Schultzer) [inline public/javascripts/receipt.js]: Use single quotation marks

File: public/css/custom.css [hunk 14]
```diff
@@ -127,3 +142,75 @@
...
+
+.receipt-window-container {
+  position: relative;
+  padding:20px;
+  text-align:center;
+  height:200px;
+}
+
+.receipt-window-container::after {
+  content: "";
```
Comment (Schultzer) [inline public/css/custom.css]: Use single quotation marks

File: public/css/custom.css [hunk 15]
```diff
@@ -127,3 +142,75 @@
     max-height: 615px;
   }
 }
+
+
+.receipt-window-container {
+  position: relative;
+  padding:20px;
+  text-align:center;
+  height:200px;
```
Comment (Schultzer) [inline public/css/custom.css]: missing space

File: public/css/custom.css [hunk 16]
```diff
@@ -127,3 +142,75 @@
     max-height: 615px;
   }
 }
+
+
+.receipt-window-container {
+  position: relative;
+  padding:20px;
+  text-align:center;
```
Comment (Schultzer) [inline public/css/custom.css]: missing space

File: public/css/custom.css [hunk 17]
```diff
@@ -40,6 +40,18 @@
   margin-top: -20px;
 }
 
+.hero .button-secondary {
+  height: auto;
+  font-size: 10px;
+  background:#fff;
```
Comment (Schultzer) [inline public/css/custom.css]: missing space

File: public/css/custom.css [hunk 18]
```diff
@@ -127,3 +142,75 @@
     max-height: 615px;
   }
 }
+
+
+.receipt-window-container {
+  position: relative;
+  padding:20px;
```
Comment (Schultzer) [inline public/css/custom.css]: missing space

File: public/css/filelist.css [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+  background-color: transparent;
+}
+
+.window .table {
+  margin: 0px;
+  padding: 0px;
+}
+
+.window .table {
+  width:100%;
```
Comment (Schultzer) [inline public/css/filelist.css]: missing space

File: public/css/filelist.css [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+}
+.window header.titlebar {
+  width:100%;
```
Comment (Schultzer) [inline public/css/filelist.css]: missing space

File: public/css/filelist.css [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+  text-align:center;
```
Comment (Schultzer) [inline public/css/filelist.css]: missing space

File: public/css/filelist.css [hunk 22]
```diff
@@ -0,0 +1,152 @@
...
+  color: rgb(60, 60, 60);
+  font-weight:400;
```
Comment (Schultzer) [inline public/css/filelist.css]: missing space

File: public/css/filelist.css [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+  font-size:12px;
```
Comment (Schultzer) [inline public/css/filelist.css]: missing space

File: public/css/filelist.css [hunk 24]
```diff
@@ -0,0 +1,152 @@
...
+  margin:0;
```
Comment (Schultzer) [inline public/css/filelist.css]: missing space

File: public/css/filelist.css [hunk 25]
```diff
@@ -0,0 +1,152 @@
...
+  justify-content: end;
+  flex-flow: row;
+}
+
+.window .cell {
+  padding: 3px 12px;
+
+  flex: 1 1 auto;
+
+  text-align:left;
```
Comment (Schultzer) [inline public/css/filelist.css]: missing space

File: public/css/filelist.css [hunk 26]
```diff
@@ -0,0 +1,152 @@
...
+
+  overflow: hidden;
+  text-overflow: ellipsis;
+  white-space: nowrap;
+}
+
+.window .cell:last-child {
+  text-align:right;
```
Comment (Schultzer) [inline public/css/filelist.css]: missing space

File: public/css/filelist.css [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+.window .head .row {
+  background-color: var(--submenu-bg);
+  border-bottom: 1px solid var(--submenu-border-buttom);
+}
+
+.window .table .body {
+  position: fixed;
+  position:absolute;
```
Comment (Schultzer) [inline public/css/filelist.css]: missing space

File: public/javascripts/receipt.js [hunk 28]
```diff
@@ -0,0 +1,140 @@
+(function(){
```
Comment (Schultzer) [inline public/javascripts/receipt.js]: missing space

File: public/javascripts/receipt.js [hunk 29]
```diff
@@ -0,0 +1,140 @@
...
+              index: 0,
+              file: {
+                name: form.querySelector('[name="document"]').files[0].name
+              },
+              done: false,
+              progressBar: 0,
+              result: {}
+            };
+          files.splice(0, 1, file);
+          var interval = setInterval(function(){
```
Comment (Schultzer) [inline public/javascripts/receipt.js]: missing space

File: public/javascripts/receipt.js [hunk 30]
```diff
@@ -0,0 +1,140 @@
...
+
+        // Don't scroll too far
+        if ((prior_position_y + distance) > destination_y)
+            distance = (prior_position_y + distance) - destination_y;
+
+        window.scrollTo(0, prior_position_y + distance);
+
+        // If we're there or can't get any closer because of the viewport's
+        // size, stop!
+        if ((window.pageYOffset >= destination_y) || (window.pageYOffset == prior_position_y)) {
```
Comment (Schultzer) [inline public/javascripts/receipt.js]: this should be `>==` and `===`

**Review by Schultzer — CHANGES_REQUESTED** [review]
So far LGTM,  it's pretty much syntax

File: controllers/api_receipt.js (hunk 3 above)
Comment (danschultzer) [reply controllers/api_receipt.js:50]: Since this is basically JSON, we should use double quotation.

**Pull request merged**